docs/principles-for-package-repository-security.md

```diff
@@ -0,0 +1,105 @@
+
+Does the package repository ...
+
+... have user accounts?
+- Yes: then you have to manage things like authentication and account recovery. For example: PyPI
+- No: For example: {index, proxy, sum}.golang.org
+
+... accept built packages, do builds on behalf of users, or only host source code?
+- Accept built packages: then you have to provide a way to link packages back to source code and build instructions. For example: npm
+- Builds on behalf of users: For example, Homebrew
+- Only host source code: For example: {index, proxy, sum}.golang.org
+
+## Security Capabilities of Package Repositories
+
+We define 4 levels of security maturity:
+- Level 0 is defined as having very little security maturity.
+- Level 1 is defined as having basic security maturity. Package repositories support basic security features like multi-factor authentication, allow security researchers to report vulnerabilities. All package management ecosystems should be working towards this level.
+- Level 2 is defined as having moderate security.
+- Level 3 is defined as having advanced security. These items are more aspirational, especially for smaller package management ecosystems.
+
+### Authentication
+
+Applies to: package repositories that have user accounts.
+
+- Level 1
+  - The package repository requires users to verify their email address. This ensures that if a user loses access to their account, they are able to recover it.
+  - The package repository documents their account recovery policy.
+  - The package repository supports multi-factor authentication (MFA) via (at minimum) TOTP.
+  - The package repository notifies maintainers via email for critical account security changes, such as password changes or disabling multi-factor authentication. This helps users detect if their account has been compromised.
+  - The package repository implements account security measures like brute force prevention (even with 2FA attempts)
+- Level 2
+  - To prevent domain resurrection for account takeover via the recovery process, the package repository detects abandoned email domains. This may look like doing a WHOIS lookup on all registered email domains, and removing the ability to recover an account via an email domain that has been abandoned.
+  - The package repository supports MFA via phishing-resistant means such as WebAuthn
+  - The package repository requires MFA for packages deemed critical (e.g. top 1% of packages by downloads).
+  - The package repository integrates with public leaked credential databases such as Have I Been Pwned to detect when users are using leaked credentials. If a user has a leaked credential, the package repository prompts them to change it upon login.
```

Comment:
I think this item could use a bit more detail, perhaps in a different, more detailed document-- if credentials have been leaked and the registry knows the user is using leaked credentials, changing the password upon login needs to, say, go through a "password reset" email, correct? Otherwise, anyone who got the leaked credentials could log in with them and change the password themselves? I suppose this goes with the item under level 1 notifying maintainers via email on password changes...

Reply:
How about this?
```suggestion
  - The package repository integrates with public leaked credential databases such as Have I Been Pwned to detect when users are using known leaked credentials upon login. If a user has a leaked credential, detected upon login, the package repository prompts them to change it. The package manager checks passwords at registration and prevents users from registering with known leaked credentials.
```